IPNI_HTTP_PROVIDER.md

```diff
@@ -0,0 +1,328 @@
+
+This endpoint retrieves the content associated with an advertisement CID and its entries. 
+The `CID` specified in the URL parameter must match the response body, as the data returned by this endpoint is immutable.
+
+To ensure proper caching and immutability, implementers must include the following response header:
+
+```
+Cache-Control: public, max-age=29030400, immutable
+```
+
+By setting this response header, it instructs client-side caches and intermediate proxies to store the response for a long duration (`max-age=29030400`), as well as treat it as immutable. This helps improve performance and ensures that the same content is served consistently for the specified `CID`.
+
+### Response
+
+The response from the `/ipni/v1/ad/{CID}` endpoint returns the data associated to the advertisement CID and its embedded links, including `Entries`. See [IPNI Specification/Advertisements](IPNI.md#advertisements) for the advertisement and entries schema. 
+
+#### Example
+
+The following represents and abbreviated advertisement encoded in `DAG-JSON`:
+
+```json
+{
+  "Addresses": [
+    "/dns4/content.example/tcp/443/wss"
+  ],
+  "ContextID": {
+    "/": {
+      "bytes": "YmFndXFlZXJh..."
+    }
+  },
+  "Entries": {
+    "/": "baguqeeraog3uqu7au2mctrfslzvua5lzl3wyoquqhbyoig6hnputypej2uvq"
+  },
+  "IsRm": false,
+  "Metadata": {
+    "/": {
+      "bytes": "gBI"
+    }
+  },
+  "PreviousID": {
+    "/": "baguqeeranqfhpdacbe7ls44ndcgeis6yufvhk4zntydlqfaplzz6mdyuu6gq"
+  },
+  "Provider": "QmQzqxhK82kAmK...",
+  "Signature": {
+    "/": {
+      "bytes": "CqsCCAASpgIwggEiMA0..."
+    }
+  }
+}
+```
+
+The following snippet shows an example of advertisement entries encoded in `DAG-JSON`: 
+
+```json
+{
+  "Entries": [
+    {
+      "/": {
+        "bytes": "EiDYa/LrNYDdHjSlKGXpHjc1IyHds9Xdi/p8d25Q5UE/bQ"
+      }
+    }
+  ]
+}
+```
+
+### Security
+
+The IPNI HTTP Provider API specification includes several security implications and considerations to ensure secure
+communication and data integrity. Here are the key security aspects:
+
+- **TLS (Transport Layer Security)**: It is recommended to use TLS whenever possible to establish secure connections
+  between clients and the IPNI HTTP provider. TLS encrypts the data during transmission, protecting it from unauthorized
+  access and tampering.
+
+- **Pagination and Maximum Content Length**: Implementers should take into account setting a maximum content length for responses to mitigate potential abuse or resource exhaustion attacks. This helps ensure the API's performance and resilience against malicious actors. We suggest a maximum content length of 4MB. It is important to note that content pagination is achieved using IPLD links. For example, advertisement entry chunks are linked together using the `Next` field, forming a linked list of multihashes. Therefore, implementors must ensure that the generated entry chunks remain below the maximum content length accepted by the indexers.
+
```

Comment:
Each chunk can be whatever size the publisher of that data decides, as long as that size and any encoding overhead is below the maximum content length. Once a chunk is created, it cannot be resized since the CID of that chunk, as represented in the advertisement or previous chunks, must be verifiable.

Reply:
If we want to have flexible chunk sizes while still being able to validate the data with the provider's signature, then the advertisement signature cannot include the first chunk CID and a separate signature is needed instead.

Reply:
Is there a specific requirement that we should satisfy which breaks the existing protocol? If not my sense is to push this out of the scope and leave this to be as is: i.e. recommendation on "consider max body sizes when chunking".

Reply:
The question is, do we want to allow the chunk size, and thus the chunk CID, to change without affecting the advertisement signature.